pkg/provider/gitlab/gitlab.go

```diff
@@ -285,14 +347,41 @@ func (g *Gitlab) Close(ctx context.Context) error {
 	return nil
 }
 
-// Validate will use the gitlab client to validate the gitlab provider using the ListVariable call to ensure get permissions without needing a specific key.
+// Validate will use the gitlab projectVariablesClient to validate the gitlab provider using the ListVariable call to ensure get permissions without needing a specific key.
 func (g *Gitlab) Validate() (esv1beta1.ValidationResult, error) {
-	_, resp, err := g.client.ListVariables(g.projectID, nil)
-	if err != nil {
-		return esv1beta1.ValidationResultError, fmt.Errorf(errList, err)
-	} else if resp == nil || resp.StatusCode != http.StatusOK {
-		return esv1beta1.ValidationResultError, fmt.Errorf(errAuth)
+	if g.projectID != "" {
+		_, resp, err := g.projectVariablesClient.ListVariables(g.projectID, nil)
+		if err != nil {
+			return esv1beta1.ValidationResultError, fmt.Errorf(errList, err)
+		} else if resp == nil || resp.StatusCode != http.StatusOK {
+			return esv1beta1.ValidationResultError, fmt.Errorf(errProjectAuth, g.projectID)
+		}
+
+		if g.inheritFromGroups {
+			projectGroups, resp, err := g.projectsClient.ListProjectsGroups(g.projectID, nil)
+			if resp.StatusCode >= 400 && err != nil {
+				return esv1beta1.ValidationResultError, err
+			}
+			discoveredIds := make([]string, len(projectGroups))
+			for i, group := range projectGroups {
+				discoveredIds[i] = strconv.Itoa(group.ID)
```

Comment:
Groups form a linked list, we need to process them in the correct order, we can not rely on the ordering of the API response.

Reply:
Thats what the ListProjectGroups call is for. It will give the linked list from most relevant to least relevant.
However I tried to find this piece in the documentation and failed.
Though I would argue that it would still be better to use that API and order the response by length of "full_path".